Sources/Compiler/Parse/Parser.swift

```diff
@@ -2465,14 +2462,15 @@ public enum Parser {
       case .name, .under:
         // function-entity-identifier
         let head = context.take()!
-        var result = SourceRepresentable(
-          value: Name(stem: String(context.lexer.source[head.range])),
-          range: head.range)
+        var labels: [String?] = []
 
         if context.currentCharacter == "(" {
-          var labels: [String?] = []
           let backup = context.backup()
           _ = context.take()
+          var closeParenFound = false
+          defer {
+            if !closeParenFound || labels.isEmpty { context.restore(from: backup) }
```

Comment:
```suggestion
            // Backtrack if we didn't find a closing parenthesis. That will let the argument-list
            // parser process the stream after the identifier to either catch a parse error if the
            // former case or parse an empty argument list in the latter.
            if !closeParenFound || labels.isEmpty { context.restore(from: backup) }
```

Note: I think errors after the opening parenthesis should be handled by the argument list parser. I expect ill-formed argument lists to be more likely than ill-formed labeled named. For example, if we see `foo(bar:???`, I think `???` the error should be diagnosed expecting `???` to be an expression, not a label.

Reply:
Don't understand your note, but thanks for the suggestion!

Reply:
@kyouko-taiga unfortunately, the suggestion doesn't address the problem, because it doesn't explain why `labels.isEmpty` should be causing backtracking.  Could you try another?

Reply:
```suggestion
            // Backtrack if we didn't find a closing parenthesis or if there are no labels. That
            // will let the argument-list parser pickup after the identifier to either catch a
            // parse error in the former case (no closing parenthesis) or parse an empty argument
            // list in the latter (no labels).
            // Note: `foo()` is *not* a valid name, it's a function call.
            if !closeParenFound || labels.isEmpty { context.restore(from: backup) }
```

Reply:
To give more context in case you want to refine the comment yourself, the rule in the specification is defined as follow

```ebnf
function-entity-identifier ::= (token)
  identifier '(' argument-label+ ')'
```

The subtlety here is that `argument-label` should appear one or more times, not zero or more. So we backtrack when we found none. That's necessary to disambiguate a function call with no argument, as suggested in the comment.